tools/codegen/gen_backend_stubs.py

```diff
@@ -44,35 +53,36 @@ def parse_backend_yaml(
 
-    metadata: Dict[OperatorName, ExternalBackendMetadata] = {}
-    for op in supported:
-        op_name = OperatorName.parse(op)
-        m = ExternalBackendMetadata(op_name, backend, is_autograd=False)
-        metadata[m.operator] = m
-    for op in supported_autograd:
-        op_name = OperatorName.parse(op)
-        m = ExternalBackendMetadata(op_name, backend, is_autograd=True)
-        metadata[m.operator] = m
-
-    native_functions_map: Dict[OperatorName, NativeFunction] = {
-        f.func.name: f
-        for f in concatMap(lambda f: [f] if isinstance(f, NativeFunction) else list(f.functions()), grouped_native_functions)
-    }
-
-    def native_to_external(
-            g: Union[NativeFunction, NativeFunctionsGroup]
-    ) -> Union[ExternalBackendFunction, ExternalBackendFunctionsGroup]:
-        if isinstance(g, NativeFunction):
-            f = g
-            m = metadata.get(f.func.name, None)
-            return ExternalBackendFunction(f, m)
-        elif isinstance(g, NativeFunctionsGroup):
-            return ExternalBackendFunctionsGroup.from_function_group(g, metadata)
+    def add_backend_index(backend_ops: List[str], backend: str, *, is_autograd: bool = False) -> DispatchKey:
+        if is_autograd:
+            # TODO: add better error messages here (later PR)
+            backend_key = DispatchKey.parse(f'Autograd{backend}')
         else:
-            assert_never(g)
-    for op_name in metadata.keys():
-        assert op_name in native_functions_map, f"Found an invalid operator name: {op_name}"
-    return cpp_namespace, [native_to_external(g) for g in grouped_native_functions]
+            backend_key = DispatchKey.parse(backend)
+        metadata: Dict[OperatorName, BackendMetadata] = {}
+        for op in backend_ops:
+            op_name = OperatorName.parse(op)
+            assert op_name in native_functions_map, f"Found an invalid operator name: {op_name}"
+            # See Note [External Backends Follow Dispatcher API]
+            kernel_name = dispatcher.name(native_functions_map[op_name].func)
+            # TODO: allow structured external backends later.
+            m = BackendMetadata(kernel=kernel_name, structured=False, external=True)
+            metadata[op_name] = m
+        assert backend_key not in backend_indices
+        # TODO: currently hardcoding the fact that XLA implements out/inplace in terms of functional ops,
+        # this should eventually be toggleable per-backend.
+        backend_indices[backend_key] = BackendIndex(dispatch_key=backend_key, use_out_as_primary=False, index=metadata)
+        return backend_key
+
+    backend_key: Optional[DispatchKey] = None
+    if len(supported) > 0:
+        backend_key = add_backend_index(supported, backend)
+
+    autograd_key: Optional[DispatchKey] = None
+    if len(supported_autograd) > 0:
+        autograd_key = add_backend_index(supported_autograd, backend, is_autograd=True)
```

Comment:
I know it's less efficient, but I probably would have preferred it if this function didn't mutate its argument and just returned a mini backend index to be merged into the main one later. (The general API design concept here is it's OK to use mutation locally, but public APIs should try to avoid mutation as much as possible, because it means that the contract for the function is a bit more complicated).

Reply:
yep good point, easy enough to perform the mutation outside of that function